### PR TITLE
Issues/6980 enhance.cache.policy

### PR DESCRIPTION
**Pull request by NicolasGeraud**
**Issue**

https://github.com/gravitee-io/issues/issues/6980

**Description**
Choose the method to cache and add a response condition

---
**Gravitee.io Automatic Deployment**

🚀 A prerelease version of this package has been published on Gravitee's private artifactory, you can:
 - use it directly by updating your project with version: `1.9.1-SNAPSHOT.issues-6980-enhance-cache-policy`
 - download it from Artifactory [here](https://odbxikk7vo-artifactory.services.clever-cloud.com/gravitee-snapshots/io/gravitee/policy/gravitee-policy-cache/1.9.1-SNAPSHOT.issues-6980-enhance-cache-policy/gravitee-policy-cache-1.9.1-SNAPSHOT.issues-6980-enhance-cache-policy.zip)

### CODE REVIEW

File: src/test/java/io/gravitee/policy/cache/CachePolicyTest.java [hunk 1]
```diff
@@ -86,9 +109,88 @@ public void should_usecachecontrol_expires_past() {
...
+
+        CachePolicy cachePolicy = new CachePolicy(cachePolicyConfiguration);
+        cachePolicy.onRequest(request, response, executionContext, policyChain);
+
+        verify(policyChain, times(1)).doNext(any(), any());
+        verify(policyChain, never()).failWith(any());
+        verify(executionContext, times(1))
+                .setAttribute(eq(ExecutionContext.ATTR_INVOKER),
+                        any(CachePolicy.CacheInvoker.class));
+    }
```
Comment (ytvnr) [inline src/test/java/io/gravitee/policy/cache/CachePolicyTest.java:195]: can we have tests with condition (passing, failing, erroring) ?

**Review by marcambier — CHANGES_REQUESTED** [review]

File: src/main/java/io/gravitee/policy/cache/proxy/EvaluableProxyResponse.java [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+    public EvaluableProxyResponse(ProxyResponse proxyResponse) {
+        super(null);
+        this.proxyResponse = proxyResponse;
+    }
+
+    public int getStatus() {
+        return proxyResponse.status();
+    }
+
+    public HttpHeaders getHeaders() {
```
[the remark this answers is not in the document]
Comment (NicolasGeraud) [reply src/main/java/io/gravitee/policy/cache/proxy/EvaluableProxyResponse.java:43]: 👍

File: src/main/java/io/gravitee/policy/cache/proxy/EvaluableProxyResponse.java [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+    public EvaluableProxyResponse(Response response) {
+        super(null);
+        throw new UnsupportedOperationException();
+    }
+    public EvaluableProxyResponse(ProxyResponse proxyResponse) {
+        super(null);
+        this.proxyResponse = proxyResponse;
+    }
+
+    public int getStatus() {
```
[the remark this answers is not in the document]
Comment (NicolasGeraud) [reply src/main/java/io/gravitee/policy/cache/proxy/EvaluableProxyResponse.java:39]: 👍

File: src/main/java/io/gravitee/policy/cache/CachePolicy.java [hunk 4]
```diff
@@ -360,4 +366,26 @@ private CacheAction lookForAction(Request request) {
...
+                    method == HttpMethod.HEAD);
+        }
+        return cachePolicyConfiguration.getMethods().contains(method);
+    }
+
+    private boolean evaluate(final ExecutionContext context, final ProxyResponse proxyResponse, final String condition) {
+        if (condition != null && !condition.isEmpty()) {
+            //context.setAttribute("proxyResponse", proxyResponse);
+            context.getTemplateEngine().getTemplateContext().setVariable(UPSTREAM_RESPONSE, new EvaluableProxyResponse(proxyResponse));
+            Boolean value = context.getTemplateEngine().getValue(condition, Boolean.class);
```
[the remark this answers is not in the document]
Comment (NicolasGeraud) [reply src/main/java/io/gravitee/policy/cache/CachePolicy.java:384]: you're right

File: src/main/java/io/gravitee/policy/cache/CachePolicy.java [hunk 5]
```diff
@@ -360,4 +366,26 @@ private CacheAction lookForAction(Request request) {
...
+            return (method == HttpMethod.GET ||
+                    method == HttpMethod.OPTIONS ||
+                    method == HttpMethod.HEAD);
+        }
+        return cachePolicyConfiguration.getMethods().contains(method);
+    }
+
+    private boolean evaluate(final ExecutionContext context, final ProxyResponse proxyResponse, final String condition) {
+        if (condition != null && !condition.isEmpty()) {
+            //context.setAttribute("proxyResponse", proxyResponse);
```
[the remark this answers is not in the document]
Comment (NicolasGeraud) [reply src/main/java/io/gravitee/policy/cache/CachePolicy.java:382]: 👍

File: src/main/java/io/gravitee/policy/cache/CachePolicy.java [hunk 6]
```diff
@@ -360,4 +366,26 @@ private CacheAction lookForAction(Request request) {
...
+            context.getTemplateEngine().getTemplateContext().setVariable(UPSTREAM_RESPONSE, new EvaluableProxyResponse(proxyResponse));
```
[the remark this answers is not in the document]
Comment (NicolasGeraud) [reply src/main/java/io/gravitee/policy/cache/CachePolicy.java:383]: I think this PR will only be focus on headers, I create another issue to support content

File: src/test/java/io/gravitee/policy/cache/CachePolicyTest.java [hunk 7]
```diff
@@ -86,9 +109,88 @@ public void should_usecachecontrol_expires_past() {
...
-        when(response.headers()).thenReturn(headers);
+        when(proxyResponse.headers()).thenReturn(headers);
 
-        long timeToLive = CachePolicy.timeToLiveFromResponse(response);
+        long timeToLive = CachePolicy.timeToLiveFromResponse(proxyResponse);
         Assert.assertEquals(-1, timeToLive);
     }
+
+    @Test
+    public void shouldFailIfNoCache() {
```
[the remark this answers is not in the document]
Comment (NicolasGeraud) [reply src/test/java/io/gravitee/policy/cache/CachePolicyTest.java:119]: 👍

File: src/main/java/io/gravitee/policy/cache/proxy/EvaluableProxyResponse.java [hunk 8]
```diff
@@ -0,0 +1,46 @@
...
+ */
+public class EvaluableProxyResponse  extends EvaluableResponse {
+    ProxyResponse proxyResponse;
+
+    public EvaluableProxyResponse(Response response) {
+        super(null);
+        throw new UnsupportedOperationException();
+    }
+    public EvaluableProxyResponse(ProxyResponse proxyResponse) {
+        super(null);
```
[the remark this answers is not in the document]
Comment (NicolasGeraud) [reply src/main/java/io/gravitee/policy/cache/proxy/EvaluableProxyResponse.java:35]: I do not support content in this PR

**Comment by NicolasGeraud** [comment]
Ok i fix this

**Review by marcambier — APPROVED** [review]

**Pull request merged**